rust-runtime/aws-smithy-client/src/retry/adaptive.rs

```diff
@@ -0,0 +1,332 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//! Adaptive Retry Strategy
+
+use crate::retry::{NewRequestPolicy, DEFAULT_MAX_ATTEMPTS};
+use aws_smithy_async::rt::sleep::AsyncSleep;
+use aws_smithy_http::operation::Operation;
+use aws_smithy_http::result::{SdkError, SdkSuccess};
+use aws_smithy_http::retry::ClassifyRetry;
+use aws_smithy_types::retry::{ErrorKind, RetryKind};
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::{Arc, Mutex};
+use std::time::Duration;
+use tokio::time::Instant;
+use tracing::Instrument;
+
+const MIN_FILL_RATE: f64 = 0.5;
+const MIN_CAPACITY: f64 = 1.0;
+const REQUEST_COST: f64 = 1.0;
+const SMOOTH: f64 = 0.8;
+/// How much to scale back after receiving a throttling response
+const BETA: f64 = 0.7;
+/// Controls how aggressively we scale up after being throttled
+const SCALE_CONSTANT: f64 = 0.4;
+
+/// Retry Policy Configuration
+///
+/// Without specific use cases, users should generally rely on the default values set
+/// by [`Config::default`](Config::default).
+///
+/// Currently these fields are private and no setters provided. As needed, this configuration
+/// will become user-modifiable in the future.
+#[derive(Clone, Debug)]
+pub struct Config {
+    max_attempts: u32,
+}
+
+impl Config {
+    /// Override the maximum number of attempts
+    ///
+    /// `max_attempts` must be set to a value of at least `1` (indicating that retries are disabled).
+    pub fn with_max_attempts(mut self, max_attempts: u32) -> Self {
+        self.max_attempts = max_attempts;
+        self
+    }
+
+    /// Returns true if retry is enabled with this config
+    pub fn has_retry(&self) -> bool {
+        self.max_attempts > 1
+    }
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            max_attempts: DEFAULT_MAX_ATTEMPTS,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Policy {
+    config: Config,
+    shared_state: CrossRequestRetryState,
+}
+
+impl Policy {
+    /// Construct a new standard retry policy from the given policy configuration.
+    pub fn new(config: Config) -> Self {
+        Self {
+            shared_state: CrossRequestRetryState::new(),
+            config,
+        }
+    }
+
+    /// Set the configuration for this retry policy.
+    pub fn with_config(&mut self, config: Config) -> &mut Self {
```

Comment:
Curious to know a use case of this method. A caller who created a `Policy` had the chance to specify `config` when calling `new` but somehow the caller wants to overwrite the initially specified `config` after that?

The same question applies to `standard::Policy::with_config`.

Reply:
Ignore this code, I didn't intend to include it in this PR. I've removed it.